### DISCUSSION TITLE
Question: Is there an existing solidity verifier for proofs created with @ethereumjs/trie?

### DISCUSSION THREAD

**Discussion by sebastiendan** [ethereumjs/ethereumjs-monorepo#2599] [Q&A]
Hi,

In combination with using `@ethereumjs/trie` in the browser to create merkle proofs against transaction tries, I'm looking for a compatible on-chain verifier. It's pretty hard to find an existing contract out there that would work out of the box with the type/format of merkle proofs created with `@ethereumjs/trie` (so many unfinished/abandoned projects).

Would you know if such a compatible on-chain verifier (contract) exist?

Thanks!

**Comment by jochem-brouwer** [comment]
This seems to me what you are looking for: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/MerkleProof.sol

**Comment by sebastiendan** [comment]
@jochem-brouwer This is for standard merkle trees only, no? I was expecting the verification of merkle proofs for merkle patricia tries to be incompatible for the one for merkle trees.

**Comment by jochem-brouwer** [comment]
This should work also for MPTs. The gotcha here is that this is a "normal" Merkle tree but the data inside the tree is encoded as Patricia tree.

**Comment by sebastiendan** [comment]
@jochem-brouwer The verifier from OpenZeppelin expects a proof as a `bytes32[]`. The proofs I get with `@ethereumjs/trie` are 307B large. Is there an extra step I'm missing?

**Comment by jochem-brouwer** [comment]
I might have been a bit too fast to reply here, but I will investigate a bit further for you, because I would expect that an EVM/Solidity implementation of MPT proofs are readily available.

**Comment by sebastiendan** [comment]
Ok thanks. In addition, we've found [this related issue](https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3886) on OpenZeppelin contract repo.

**Comment by jochem-brouwer** [comment]
(have converted to a discussion :) )

**Comment by sebastiendan** [comment]
@jochem-brouwer Any luck so far with your research? So far we're only finding verifier contracts that work with specific formats expected from provers.

**Comment by sebastiendan** [comment]
FYI we were able to use this Solidity verifier to verify Merkle proofs of transaction tries made with @ethereumjs/trie:
https://github.com/lorenzb/proveth

We had to not use `trie.createProof` which is essentially finding the path in the trie and serializing the nodes in the path, and rather use the `trie.findPath` function along with a `node => node.raw()` map.

**Comment by okwme** [comment]
would love to have this functionality. Seems the discussion has been cold over a year.
